### PR TITLE
Fix batch_outputs with optimizer frequencies

### LINKED ISSUE #3143: Trainer crashed when optimizer frequency is defined.
### To Reproduce

Run the following code:
https://gist.github.com/24hours/ec67de5384bb05e28544d580ae424639

```
Traceback (most recent call last):
  File "pl_bug.py", line 40, in <module>
    trainer.fit(mnist_model, train_loader) 
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/states.py", line 48, in wrapped_fn
    result = fn(self, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 1073, in fit
    results = self.accelerator_backend.train(model)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/accelerators/gpu_backend.py", line 51, in train
    results = self.trainer.run_pretrain_routine(model)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 1239, in run_pretrain_routine
    self.train()
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 394, in train
    self.run_training_epoch()
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 491, in run_training_epoch
    batch_output = self.run_training_batch(batch, batch_idx)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 883, in run_training_batch
    batch_outputs[opt_idx].append(opt_closure_result.training_step_output_for_epoch_end)
```

#### Code sample

```python
    def configure_optimizers(self):
        optimizer_G = torch.optim.Adam(self.parameters(), lr=0.1, weight_decay=1e-5)
        optimizer_D = torch.optim.Adam(self.parameters(), lr=0.1, weight_decay=1e-5)

        return [
                {'optimizer': optimizer_D, 'frequency': 5},
                {'optimizer': optimizer_G, 'frequency': 1}
            ]
```

the culprit is 'frequency' : 5, removing the line will allow trainer to run smoothly. 

https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html?highlight=optimizers#configure-optimizers
The definition is correct according to this documentation. 

### Expected behavior

Model should train without crash. 
The code work in 0.8.5 environment. 

### Environment

* CUDA:
	- GPU:
		- GeForce GTX TITAN X
	- available:         True
	- version:           11.0
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0a0+9907a3e
	- pytorch-lightning: 0.9.0
	- tensorboard:       2.2.0
	- tqdm:              4.48.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.10
	- version:           #110-Ubuntu SMP Tue Jun 23 02:39:32 UTC 2020

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #3143

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/base/model_optimizers.py [hunk 1]
```diff
@@ -34,6 +34,14 @@ def configure_optimizers__multiple_optimizers(self):
         optimizer2 = optim.Adam(self.parameters(), lr=self.learning_rate)
         return optimizer1, optimizer2
 
+    def configure_optimizers__multiple_optimizers_frequency(self):
```
Comment (Borda) [inline tests/base/model_optimizers.py:37]:
```suggestion
    def configure_optimizers___multiple_optimizers_frequency(self):
```
pls check the other syntax, but I gues it shall be 3x "_"

Comment (rohitgr7) [reply tests/base/model_optimizers.py:37]: it's 2 everywhere.

**Comment by rohitgr7** [comment]
ping @williamFalcon.

**Comment by rohitgr7** [comment]
open to suggestions for tests if required here :)

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -870,7 +870,10 @@ def run_training_batch(self, batch, batch_idx):
                 self.batch_loss_value.append(opt_closure_result.loss)
 
                 # track all the outputs across all steps
-                batch_outputs[opt_idx].append(opt_closure_result.training_step_output_for_epoch_end)
+                if len(batch_outputs) == 1:
+                    batch_outputs[0].append(opt_closure_result.training_step_output_for_epoch_end)
+                else:
+                    batch_outputs[opt_idx].append(opt_closure_result.training_step_output_for_epoch_end)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:876]:
```suggestion
                idx = opt_idx if len(batch_outputs) > 1 else 0
                batch_outputs[idx].append(opt_closure_result.training_step_output_for_epoch_end)
```
maybe find better name for "idx"

**Review by awaelchli — APPROVED** [review]
looks good, test fails on master. 
this seems an important fix, I would add note to changelog

**Review by SkafteNicki — APPROVED** [review]

File: docs/source/converting.rst [hunk 3]
```diff
@@ -54,9 +54,9 @@ Lightning automates most of the trining for you, the epoch and batch iterations,
 ===========================
 To add an (optional) validation loop add logic to :func:`pytorch_lightning.core.LightningModule.validation_step` hook (make sure to use the hook parameters, self in this case).
 
-.. testcode::
+.. code-block::
```
Comment (Borda) [inline docs/source/converting.rst:57]: what was wrong with this test?

Comment (rohitgr7) [reply docs/source/converting.rst:57]: it was failing on master since pl is not defined here. Also all other blocks are code-blocks up there.

Comment (Borda) [reply docs/source/converting.rst:57]: You can use it without pl...
Cc: @awaelchli

Comment (rohitgr7) [reply docs/source/converting.rst]: ok, fixed

**Comment by rohitgr7** [comment]
@Borda this one too?

File: CHANGELOG.md [hunk 4]
```diff
@@ -30,6 +30,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed `GpuUsageLogger` to work on different platforms ([#3008](https://github.com/PyTorchLightning/pytorch-lightning/pull/3008))
 
+- Fixed batch_outputs with optimizer frequencies ([#3229](https://github.com/PyTorchLightning/pytorch-lightning/pull/3229))
```
Comment (Borda) [inline CHANGELOG.md:33]:
```suggestion
- Fixed `batch_outputs` with optimizer frequencies ([#3229](https://github.com/PyTorchLightning/pytorch-lightning/pull/3229))
```

**Review by Borda — APPROVED** [review]
LGTM :]

**Comment by awaelchli** [comment]
@rohitgr7 this touches the training loop and @williamFalcon is currently refactoring that. It looks like the training loop will move a separate class (see training_loop_temp.py), so we need to be careful here and coordinate with him.

**Comment by rohitgr7** [comment]
@awaelchli the core training loop is not implemented yet in that file. So I guess either we should block it for some time or merge it now since I don't think the core loop will be changed.

**Comment by awaelchli** [comment]
agreed

**Comment by rohitgr7** [comment]
@Borda can we merge this?

**Pull request merged**